web/packages/teleterm/src/mainProcess/agentDownloader/fileDownloader.ts

```diff
@@ -0,0 +1,124 @@
+
+import Logger from 'teleterm/logger';
+
+export interface IFileDownloader {
+  run(url: string, downloadDirectory: string): Promise<void>;
+}
+
+export class FileDownloader implements IFileDownloader {
+  private logger = new Logger('fileDownloader');
+  private onResolve: () => void;
+  private onReject: (error: Error) => void;
+
+  constructor(private window: BrowserWindow) {}
+
+  async run(url: string, downloadDirectory: string) {
+    this.logger.info(
+      `Starting download from ${url} (download directory: ${downloadDirectory}).`
+    );
+    const handler = this.createDownloadHandler(url, downloadDirectory);
+    this.window.webContents.session.on('will-download', handler);
+
+    try {
+      return await new Promise<void>((resolve, reject) => {
+        this.onResolve = resolve;
+        this.onReject = reject;
+        this.window.webContents.downloadURL(url);
+      });
+    } finally {
+      this.window.webContents.session.off('will-download', handler);
+    }
+  }
+
+  private createDownloadHandler(url: string, downloadDirectory: string) {
+    return (_: Event, item: DownloadItem) => {
+      const isExpectedUrl = item.getURL() === url;
+      if (!isExpectedUrl) {
+        // handle only expected URL
+        return;
+      }
+
+      item.on('updated', (_, state) => {
+        switch (state) {
+          case 'interrupted':
+            this.onDownloadError(new Error('Download failed: interrupted'));
+            break;
+          case 'progressing':
+            this.onProgress(item.getReceivedBytes(), item.getTotalBytes());
+        }
+      });
+
+      item.once('done', (_, state) => {
+        switch (state) {
+          case 'completed':
+            this.onDownloadDone();
+            break;
+          case 'interrupted':
+            // TODO(gzdunek): electron doesn't expose much information about why the download failed.
+            // Fortunately, there is a PR in works that will add more info https://github.com/electron/electron/pull/38859.
+            // Use Use DownloadItem.getLastReason() when merged.
+            this.onDownloadError(
+              new Error(
+                `Download failed. Requested file may not exist or is temporary unavailable.`
+              )
+            );
+            break;
+          case 'cancelled':
+            // We cancel the download only when filename is incorrect, so we can mention that in the error message.
+            this.onDownloadError(
+              new Error(
+                `Download was cancelled. Filename ${item.getFilename()} is incorrect.`
+              )
+            );
+        }
+      });
+
+      // The filename is taken from Content-Disposition header, so it is worth to check if it does not contain
+      // any disallowed characters that could cause storing the file outside the download directory.
+      if (path.basename(item.getFilename()) !== item.getFilename()) {
+        item.cancel();
```

Comment:
Is there a guideline that says we should do this? I'd expect Chromium to handle that.

Reply:
Okay, I tested that.
I was slightly incorrect about headers that come from our CDN. We don't even have `Content-Disposition` set, but only `Content-Type: application/octet-stream`. In that case, the name is taken from URL. Because of that, you can't simply add `../` to the filename, as it just makes the URL to point to a  different file on the server. In our case it will cause a file selector to show up, because we check if `item.getURL() === url`. That won't pass and the default handler would handle the file right from the start.

I also tried providing incorrect filenames and Chromium seems to escape them:
- `http://localhost:7001/..%2Fabcd` -> filename `_abcd`
-  `http://localhost:7001/abcd:` -> filename `abcd_`

I also tried passing the filename in the `Content-Disposition` header:
- `Content-Disposition': "attachment; filename='../MyFileName.ext'` -> filename `.._MyFileName.ext`

I will remove our custom check :)